src/lsst/cmservice/common/graph.py

```diff
@@ -344,10 +346,56 @@ async def append_node_to_graph(
             s = select(Edge).with_for_update().where(Edge.source == node_0)
             downstream_edges = (await session.exec(s)).all()
         case ManifestKind.grouped_step:
-            # node_1 should become adjacent to the next node in node_0's path
-            # that is not a group or a collect node.
+            # node_1 should become adjacent to the next nodes in node_0's path
+            # that are not a group or a collect node.
+            # this ORM statement forms a recursive CTE for traversing a graph
+            # starting at the edge for which node_0 is a source. The recursive
+            # CTE returns all edges along the directed path from node_0, and
+            # in the end we ignore edges involving nodes that are "invisible"
+            # to this operation. The general form of this Query is
+            #
+            # WITH RECURSIVE ed AS (
+            #  SELECT * from edges where source = node_0
+            #  UNION
+            #  SELECT * from edges JOIN ed on (edges.source = ed.target)
+            # ) SELECT * from ed
+            # join nodes on (edge.target = node.id)
+            # where node.kind not in (...)
+            cte = (
+                select(
+                    col(Edge.id).label("id"),
+                    col(Edge.source).label("source"),
+                    col(Edge.target).label("target"),
+                )
+                .where(Edge.source == node_0)
+                .cte("ed", recursive=True)
+            )
+            next_valid_nodes = select(Edge.id, Edge.source, Edge.target).join(
+                cte, col(Edge.source) == cte.c.target
+            )
+            s = (
+                select(Edge)
+                .join(cte.union(next_valid_nodes), col(Edge.id) == cte.c.id)
+                .join(Node, col(Edge.target) == Node.id)
+                .where(col(Node.kind).not_in([ManifestKind.group, ManifestKind.collect_groups]))
+                .limit(1)
```

Comment:
Just trying to think of any ways this could behave unexpectedly: say we had two steps and tried to append to groups in the first step, is this not going to walk the whole graph of edges and potentially pick the wrong non-group, non-collect downstream node, from the second step? (i.e. if the limit was > 1, what else would be returned?)